File: Algorithm/src/twopointers/C1_M_4Sum/C1_M_4Sum.java
package twopointers.C1_M_4Sum;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class C1_M_4Sum {
  public static List<List<Integer>> quadruplets(int[] arr, int target){
    Arrays.sort(arr);
    List<List<Integer>> res = new ArrayList<>();
    for(int i = 0; i < arr.length - 3; i++){
      if(i > 0 && arr[i] == arr[i - 1])
        continue;
      for(int j = i + 1; j < arr.length - 2; j++){
        if(j > i + 1 && arr[j] == arr[j - 1])
          continue;
        pair(arr, target, i, j, res);
      }
    }
    return res;
  }

  public static void pair(int[] arr, int target, int first, int second, List<List<Integer>> res){
    int left = second + 1;
    int right = arr.length - 1;
    while(left < right){
      int sum = arr[first] + arr[second] + arr[left] + arr[right];
      if(sum == target){
        res.add(Arrays.asList(arr[first], arr[second], arr[left], arr[right]));
        left++;
        right--;
        while(left < right && arr[left] == arr[left - 1]) left++;
        while(left < right && arr[right] == arr[right + 1]) right--;
      }else if(sum > target){
        right--;
      }else{
        left++;
      }
    }
  }

  public static void main(String[] args) {
    System.out.println(quadruplets(new int[] {4,1,2,-1,1,-3}, 1));
  }
}

File: Algorithm/src/twopointers/C1_M_4Sum/C1_M_4Sum.py
def quadruplets(arr, target):
  arr.sort()
  res = []
  for i in range(len(arr) - 3):
    if i > 0 and arr[i] == arr[i - 1]:
      continue
    for j in range(i + 1, len(arr) - 2):
      if j > i + 1 and arr[j] == arr[j - 1]:
        continue
      pair(arr, target, i, j, res)
  return res

def pair(arr, target, first, second, res):
  left = second + 1
  right = len(arr) - 1
  while left < right:
    sum = arr[first] + arr[second] + arr[left] + arr[right]
    if sum == target:
      res.append([arr[first], arr[second], arr[left], arr[right]])
      left += 1
      right -= 1
      while left < right and arr[left] == arr[left - 1]:
        left += 1
      while left < right and arr[right] == arr[right + 1]:
        right -= 1
    elif sum > target:
      right -= 1
    else:
      left += 1

def main():
  print(quadruplets([4,1,2,-1,1,-3], 1))

main()